community/kernel/src/main/java/org/neo4j/kernel/configuration/ssl/SslPolicyConfig.java

```diff
@@ -0,0 +1,112 @@
+import static org.neo4j.kernel.configuration.Settings.PATH;
+import static org.neo4j.kernel.configuration.Settings.STRING;
+import static org.neo4j.kernel.configuration.Settings.STRING_LIST;
+import static org.neo4j.kernel.configuration.Settings.derivedSetting;
+import static org.neo4j.kernel.configuration.Settings.options;
+import static org.neo4j.kernel.configuration.Settings.pathSetting;
+import static org.neo4j.kernel.configuration.Settings.setting;
+
+@Group( "dbms.ssl.policy" )
+public class SslPolicyConfig
+{
+    private static final String TLS_VERSION_DEFAULTS = join( ",", new String[]{"TLSv1", "TLSv1.1", "TLSv1.2"} );
+
+    @Description( "The mandatory base directory for cryptographic objects of this policy." +
+                  " It is also possible to override each individual configuration with absolute paths." )
+    public final Setting<File> base_directory;
+
+    @Description( "Allows the generation of a private key and associated self-signed certificate." +
+                  " Only performed when both objects cannot be found." )
+    public final Setting<Boolean> allow_key_generation;
+
+    @Description( "Makes this policy trust all remote parties." +
+                  " Enabling this is not recommended and the trusted directory will be ignored." )
+    public final Setting<Boolean> trust_all;
+
+    @Description( "Private PKCS#8 key in PEM format." )
+    public final Setting<File> private_key;
+
+    @Internal // not yet implemented
+    @Description( "The password for the private key." )
+    public final Setting<String> private_key_password;
+
+    @Description( "X.509 certificate (chain) of this server in PEM format." )
+    public final Setting<File> public_certificate;
+
+    @Description( "Path to directory of X.509 certificates in PEM format for trusted parties." )
+    public final Setting<File> trusted_dir;
+
+    @Internal // not yet implemented
+    @Description( "Path to directory of CRLs." )
+    public final Setting<File> revoked_dir;
+
+    @Description( "Client authentication stance." )
+    public final Setting<ClientAuth> client_auth;
+
+    @Description( "Restrict allowed TLS protocol versions." )
+    public final Setting<List<String>> tls_versions;
+
+    @Description( "Restrict allowed ciphers." )
+    public final Setting<List<String>> ciphers;
+
+    public SslPolicyConfig( String policyName )
+    {
+        GroupSettingSupport group = new GroupSettingSupport( SslPolicyConfig.class, policyName );
+
+        this.base_directory = group.scope( pathSetting( "base_directory", NO_DEFAULT ) );
+        this.allow_key_generation = group.scope( setting( "allow_key_generation", BOOLEAN, FALSE ) );
+        this.trust_all = group.scope( setting( "trust_all", BOOLEAN, FALSE ) );
+
+        this.private_key = group.scope( derivedDefault( "private_key", base_directory, "private.key" ) );
+        this.public_certificate = group.scope( derivedDefault( "public_certificate", base_directory, "public.crt" ) );
```

Comment:
1. I think it is fine to merge this with TODOs in place. It is functionally complete for the functionality that is in. The TODOs are clean cut-off points.

2. This is mainly an exercise in integration and refactoring so most tests stay in place, for example for TLS support of Bolt and webserver. They work the same under the legacy policy. The new functionality with regards to secure communication does have tests and ITs.

3. Me neither. That is what the licensing tool did so I will have to assume it is as it should be. I will try to double-check with someone else though.